src/lfortran/semantics/ast_body_visitor.cpp

```
@@ -3218,6 +3218,22 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
         Vec<ASR::stmt_t*> orelse;
         orelse.reserve(al, x.n_orelse);
         transform_stmts(orelse, x.n_orelse, x.m_orelse);
+        // check if `test` is stricly an array and not the ttype of an accepted expression
+        if (ASRUtils::is_array(ASRUtils::expr_type(test))
+            && !ASR::is_a<ASR::IntegerCompare_t>(*test) 
+            && !ASR::is_a<ASR::RealCompare_t>(*test)
+            && !ASR::is_a<ASR::LogicalBinOp_t>(*test)) {
+            if (!ASR::is_a<ASR::Logical_t>(
+                    *ASRUtils::type_get_past_array(ASRUtils::expr_type((test))))) {
+                throw SemanticError("'where' clause requires a logical array", test->base.loc);
+            } else {
+                ASR::expr_t* logical_true = ASRUtils::EXPR(
+                                                ASR::make_LogicalConstant_t(al, x.base.base.loc, true,
+                                                ASRUtils::TYPE(ASR::make_Logical_t(al, x.base.base.loc, 4))));
+                test = ASRUtils::EXPR(ASR::make_LogicalBinOp_t(al, x.base.base.loc, test,
+                            ASR::logicalbinopType::Eqv, logical_true, ASRUtils::expr_type(test), nullptr));
+            }
+        }
```

The logic doesn't seem right. Let's say I have a complex array. Then it seems it will go into the "else" branch here, and tries to compare it to logical True? That seems wrong.

Also, what if it is not an array --- shouldn't it fail then?

Instead, why not compare the type itself of the array element? 

I like to do the non-failing mode first in these nested ifs, like this:

```
if (ASRUtils::is_array(ASRUtils::expr_type(test)) {
    if (array_element_ttype(test) == logical) {
        if (logical op) {
            // already in the right format (maybe later we can insert more nested ifs here with more checks)
        } else {
            // explicitly compare against True
        }
    } else {
        error "the first array argument to `where` must be of type logical"
    }
} else {
    error "the argument to `where` must be an array"
}
```

---

> Let's say I have a complex array

I think we cannot have it, it should throw an error. @kmr-srbh can you please check and revert if gfortran throws an error or not. By the change you did, we will throw an error for sure.

---

GFortran throws an error and we do so too. We test it here - https://github.com/lfortran/lfortran/pull/4320/files#diff-a07067ee3d9cce4783c7a91018d10ef98a00c204321fed16955a53b9e81e96cf.

---

In latest main we have no type checking for `test`, with this PR as we are handling types, we shall do it in non-failing mode as Ondrej suggested above.

---

> should necessarily always be a logical array.

You are correct, this is what we expect it to be, but end user may write code like `where(.true.)` and in this case type of test will be a logical scalar and it is when we need to throw an error. We cannot certainly say that user will write code correctly, but we need to be certain that we pass a LogicalArray to our ASR passes to make its implementation simple.

---

You may refer code block present in https://github.com/lfortran/lfortran/pull/4320/files#r1647528202 and try to understand how your current implementation and that differs.